File: frontend/modules/velo-download-engine/android/src/main/java/expo/modules/velodownloadengine/VeloMediaToolsModule.kt
package expo.modules.velodownloadengine

import android.content.Context
import android.net.Uri
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/** JS-facing wrapper for [MediaTools]. Engine errors surface to JS with their code (e.g. FORMAT_UNAVAILABLE). */
class VeloMediaToolsModule : Module() {
  private val ctx: Context get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  private fun <T> guarded(block: () -> T): T = try {
    block()
  } catch (e: EngineError) {
    throw CodedException(e.code, e.message, e)
  } catch (e: DownloadCancelled) {
    throw CodedException("CANCELLED", "Cancelled", e)
  } catch (e: SecurityException) {
    throw CodedException("MEDIA_NOT_FOUND", "No permission to read this file", e)
  }

  override fun definition() = ModuleDefinition {
    Name("VeloMediaTools")
    Events("onToolProgress")

    fun progress(opId: String) = { p: Double -> sendEvent("onToolProgress", mapOf("opId" to opId, "progress" to p)) }

    AsyncFunction("probe") { uri: String -> guarded { MediaTools.probe(ctx, Uri.parse(uri)) } }
    AsyncFunction("extractAudio") { uri: String, filename: String, relPath: String, opId: String ->
      guarded { MediaTools.extractAudio(ctx, Uri.parse(uri), filename, relPath, opId, progress(opId)) }
    }
    AsyncFunction("trim") { uri: String, startMs: Double, endMs: Double, filename: String, relPath: String, opId: String ->
      guarded { MediaTools.trim(ctx, Uri.parse(uri), startMs.toLong(), endMs.toLong(), filename, relPath, opId, progress(opId)) }
    }
    AsyncFunction("frame") { uri: String, atMs: Double, format: String, filename: String, relPath: String ->
      guarded { MediaTools.frame(ctx, Uri.parse(uri), atMs.toLong(), format, filename, relPath) }
    }
    AsyncFunction("imageProcess") { uri: String, format: String, quality: Int, maxDimension: Int, filename: String, relPath: String ->
      guarded { MediaTools.imageProcess(ctx, Uri.parse(uri), format, quality, maxDimension, filename, relPath) }
    }
    AsyncFunction("sha256") { uri: String, opId: String -> guarded { MediaTools.sha256(ctx, Uri.parse(uri), opId) } }
    AsyncFunction("cancel") { opId: String -> MediaTools.cancel(opId) }

    AsyncFunction("describe") { uri: String -> MediaStoreWriter.describe(ctx, Uri.parse(uri)) }
    AsyncFunction("describeMany") { uris: List<String> -> uris.map { MediaStoreWriter.describe(ctx, Uri.parse(it)) } }
    AsyncFunction("move") { uri: String, relPath: String, name: String, kind: String -> MediaStoreWriter.move(ctx, Uri.parse(uri), relPath, name, kind) }
  }
}

File: frontend/modules/velo-download-engine/src/VeloMediaToolsModule.ts
import { NativeModule, requireNativeModule } from 'expo';

import type { MediaEntry, MediaProbe, MediaToolsEvents, ToolResult } from './VeloDownloadEngine.types';

declare class VeloMediaToolsModule extends NativeModule<MediaToolsEvents> {
  probe(uri: string): Promise<MediaProbe>;
  extractAudio(uri: string, filename: string, relPath: string, opId: string): Promise<ToolResult>;
  trim(uri: string, startMs: number, endMs: number, filename: string, relPath: string, opId: string): Promise<ToolResult & { startMs: number; kind: 'video' | 'audio' }>;
  frame(uri: string, atMs: number, format: 'jpg' | 'png', filename: string, relPath: string): Promise<ToolResult>;
  imageProcess(uri: string, format: 'jpg' | 'png' | 'webp', quality: number, maxDimension: number, filename: string, relPath: string): Promise<ToolResult & { width: number; height: number }>;
  sha256(uri: string, opId: string): Promise<string>;
  cancel(opId: string): Promise<void>;
  describe(uri: string): Promise<MediaEntry | null>;
  describeMany(uris: string[]): Promise<(MediaEntry | null)[]>;
  move(uri: string, relPath: string, name: string, kind: string): Promise<boolean>;
}

export default requireNativeModule<VeloMediaToolsModule>('VeloMediaTools');
